<!DOCTYPE html>
<html lang="en">
<head>
<title>ComplexHeatmap</title>

<link rel="stylesheet" href="assets/css/splitter.css">    
<link rel="stylesheet" href="assets/css/w3.css">    
<link rel="stylesheet" href="assets/css/loading-overlay.css">
<style type="text/css" media="screen">
	iframe { 
	  z-index: 999999;
	}
</style>
<script src="assets/js/underscore-min.js"></script>    
<script src='assets/js/jquery.min.js'></script>
<script src='assets/js/jquery-resizable.js'></script>
<script src='assets/js/loading-overlay.min.js'></script>

<script src="opencpu-0.5.js"></script>
<script src="opencpu-server.js"></script>
<script src="src-min-noconflict/ace.js" type="text/javascript" charset="utf-8"></script>

<script>
	var editor;
	var resultHTML;
	
	function domarkdown(e){
       	$('#resultdiv').loadingOverlay();
		$('#runbtn').prop('disabled', true);

		var req = ocpu.call("rmdtext", {
			text : editor.getSession().getValue()
		}, function(session){
			console.log(session);
			resultHTML = session.getFileURL("output.html");
			$("iframe").attr('src', resultHTML);

			if ($('#resultdiv').hasClass('loading')) {
	        	$('#resultdiv').loadingOverlay('remove');
	      	}
			$('#runbtn').prop('disabled', false);
		}).fail(function(text){
			alert("Error: " + req.responseText);
			if ($('#resultdiv').hasClass('loading')) {
	        	$('#resultdiv').loadingOverlay('remove');
	      	}
			$('#runbtn').prop('disabled', false);
		});
	}
	
	$(function(){
		editor = ace.edit("coding");
		editor.setTheme("ace/theme/iplastic");
		editor.getSession().setMode("ace/mode/r");
		editor.getSession().setUseWrapMode(true);
		editor.setFontSize("14px");
	});
</script>
</head>

<body>
&nbsp;<button id="runbtn" class="w3-button w3-green w3-hover-white w3-hover-text-green" style="font-size:16px" onclick="domarkdown();">Run &raquo;</button>
&nbsp;<button class="w3-button w3-bar-item w3-green w3-hover-white w3-hover-text-green" style="font-size:16px" onclick="winf=window.open(resultHTML); return false;">Print</button>
<div class="panel-container">
	<div id=coding class="panel-left">

---
output:
  html_document:
    code_folding: hide
---

## [ComplexHeatmap Reference Book](https://jokergoo.github.io/ComplexHeatmap-reference/book/)
***

```{r, message=FALSE, fig.width=12, fig.height=12, dev='svg', class.source = 'fold-show'}
library(ComplexHeatmap)
library(circlize)
```

# Figure 1
```{r, message=FALSE, fig.width=12, fig.height=12, dev='svg'}

set.seed(1)
mat <- matrix(runif(36), nrow=6, dimnames=list(paste("row", letters[1:6], sep="_"), paste("col", LETTERS[1:6], sep="_")))

ha_mix_top <- HeatmapAnnotation(
  points = anno_points(runif(ncol(mat)), height = unit(2, "cm")), ## points
  barplot = anno_barplot(runif(ncol(mat)), axis=TRUE, height = unit(2, "cm")), ## barplot
  histogram = anno_histogram(mat, gp = gpar(fill = 1:6)), ## histogram
  density_line = anno_density(mat, type = "line", gp = gpar(col = 1:6)), ## density
  violin = anno_density(mat, type = "violin", gp = gpar(fill = 6:1)), ## violin
  box = anno_boxplot(mat), ## boxplot
  heatmap = anno_density(mat, type = "heatmap"), ## heatmap
  height = unit(10, "cm")
)

ha1 <- rowAnnotation(b1 = row_anno_boxplot(mat, axis = TRUE), p1 = row_anno_points(runif(nrow(mat)), axis = TRUE), width = unit(4, "cm"))
ha2 <- rowAnnotation(b2 = row_anno_boxplot(mat, axis = TRUE), p2 = row_anno_points(rowMeans(mat), axis = TRUE), width = unit(4, "cm"))

Heatmap(mat, name = "foo", top_annotation = ha_mix_top) + ha1 + ha2

decorate_annotation("barplot", { grid.text("barplot", unit(-8, "mm"), just="bottom", rot=90)})
decorate_annotation("barplot", { grid.lines(c(0, 1), unit(c(.5, .5), "native"), gp=gpar(lty=2, col="red")) })
decorate_annotation("p2", { grid.lines(unit(c(.5, .5), "native"), c(0, 1), gp=gpar(col="red")) })
```

# Figure 2

```{r, message=FALSE, fig.width=12, fig.height=8, dev='svg'}

expr = readRDS(system.file(package = "ComplexHeatmap", "extdata", "gene_expression.rds"))
mat = as.matrix(expr[, grep("cell", colnames(expr))])
base_mean = rowMeans(mat)
mat_scaled = t(apply(mat, 1, scale))

type = gsub("s\\d+_", "", colnames(mat))
ha = HeatmapAnnotation(type = type, annotation_name_side = "left")

ht_list = Heatmap(mat_scaled, name = "expression", row_km = 5, 
    col = colorRamp2(c(-2, 0, 2), c("green", "white", "red")),
    top_annotation = ha, 
    show_column_names = FALSE, row_title = NULL, show_row_dend = FALSE) +
Heatmap(base_mean, name = "base mean", 
    top_annotation = HeatmapAnnotation(summary = anno_summary(gp = gpar(fill = 2:6), 
        height = unit(2, "cm"))),
    width = unit(15, "mm")) +
rowAnnotation(length = anno_points(expr$length, pch = 16, size = unit(1, "mm"), 
    axis_param = list(at = c(0, 2e5, 4e5, 6e5), 
        labels = c("0kb", "200kb", "400kb", "600kb")),
    width = unit(2, "cm"))) +
Heatmap(expr$type, name = "gene type", 
    top_annotation = HeatmapAnnotation(summary = anno_summary(height = unit(2, "cm"))), width = unit(15, "mm"))

ht_list = rowAnnotation(block = anno_block(gp = gpar(fill = 2:6, col = NA)), 
    width = unit(2, "mm")) + ht_list

draw(ht_list, ht_gap = unit(5, "mm"))
```

# Figure 3

```{r, message=FALSE, fig.width=12, fig.height=8, dev='svg'}
df <- scale(mtcars)
# Define some graphics to display the distribution of columns
.hist = anno_histogram(df, gp = gpar(fill = "lightblue"))
.density = anno_density(df, type = "line", gp = gpar(col = "blue"))
ha_mix_top = HeatmapAnnotation(hist = .hist, density = .density, height = unit(3.8, "cm"))

# Define some graphics to display the distribution of rows
.violin = anno_density(df, type = "violin", gp = gpar(fill = "lightblue"), which = "row")
.boxplot = anno_boxplot(df, which = "row")

ha_mix_right = HeatmapAnnotation(violin = .violin, bxplt = .boxplot, which = "row", width = unit(4, "cm"))
# Combine annotation with heatmap
Heatmap(df, name = "mtcars", column_names_gp = gpar(fontsize = 8), top_annotation = ha_mix_top) + ha_mix_right
```


# Figure 4
```{r, message=FALSE, fig.width=12, fig.height=8, dev='svg'}
samples=c(paste0("X",c(4:7,10,14),"d"), paste0 ("X",c(0,1,3,6,9,12,24,36,48,60,72),"h"))
df3=data.frame(replicate (17,rnorm(17,5,1)), row.names = samples)
colnames(df3)=samples

df3rowann <-
    HeatmapAnnotation(
        violin = anno_density(
            data.matrix(df3),
            type = "line",
            gp = gpar(fill = rep(c('#00944C','#DD7931'), c(6, 11))),
            which = "row"
        ),
        boxplot = anno_boxplot(
            data.matrix(df3),
            border = TRUE,
            gp = gpar(fill = rep(c('#00944C','#DD7931'), c(6, 11))),
            pch = ".",
            axis = TRUE,
            axis_param = "bottom",
            axis_axis = gpar(fontsize = 12),
            which = "row"
        ),
        which = "row",
        width = unit(8, "cm")
    )
## Create column annotations.
df3topCol <-
    HeatmapAnnotation(
            violin = anno_density(
                data.matrix(df3),
                type = "violin",
                gp = gpar(fill = rep(c('#00944C','#DD7931'),c(6,11)))),
            boxplot = anno_boxplot(
            data.matrix(df3),
            border = TRUE,
            gp = gpar(fill = rep(c('#00944C','#DD7931'),c(6,11))),
            pch = ".",
            axis = TRUE,
            axis_param = "left",
            axis_axis = gpar(fontsize = 12),
        ),
        which = "col",
		sex = rep(c('M','F'),c(6,11)),
        col = list(sex=c("M"='#00944C',"F"='#DD7931')),
		height = unit(8, "cm")
    )
## Main plot
df3rowann + Heatmap(
    df3,
    cell_fun = function(j,i,x, y, width, height, fill) {
        grid.text(sprintf("%.f", df3[i,j]), x, y, gp = gpar(fontsize = 14))
    },
    cluster_rows = FALSE,
    cluster_columns = FALSE,
    top_annotation = df3topCol
)
```

# Figure 5
```{r, message=FALSE, fig.width=12, fig.height=8, dev='svg'}

lt = readRDS(system.file("extdata", package = "ComplexHeatmap", "dmr_summary.rds"))

# Attach all these variables to the working environment.
attach(lt)

meth_col_fun = colorRamp2(c(0, 0.5, 1), c("blue", "white", "red"))
corr_col = c("green", "red")
dist_tss_col = c("#FF0000", "#FF7352", "#FFB299", "#FFD9CB")
gene_anno_col = c("green", "blue")
cgi_anno_col = c("#FFA500", "#FFD191")
z_score_col_fun = colorRamp2(c(-200, 0, 200), c("green", "white", "red"))
state_col = c("#FF0000", "#008000", "#C2E105", "#8A91D0", "#CD5C5C", "#808080", "#000000")

anno_width = unit(3, "cm")
ht_list = rowAnnotation(text = anno_text(label, location = unit(1, "npc"), just = "right", 
    gp = gpar(fontsize = 12)))

ht_list = ht_list + Heatmap(mean_meth, name = "mean_meth", col = meth_col_fun, 
    cluster_rows = FALSE, row_title = NULL, cluster_columns = FALSE, show_row_names = FALSE,
    heatmap_legend_param = list(title = "Methylation"), width = ncol(mean_meth)*unit(4, "mm")) +
rowAnnotation("n_gr" = anno_barplot(n_gr, bar_width = 1, width = anno_width), 
    show_annotation_name = FALSE) +
rowAnnotation("n_corr" = anno_barplot(n_corr, bar_width = 1, gp = gpar(fill = corr_col), 
    width = anno_width), show_annotation_name = FALSE) +
rowAnnotation("dist_tss" = anno_barplot(dist_tss, bar_width = 1, gp = gpar(fill = dist_tss_col), 
    width = anno_width), show_annotation_name = FALSE) +
rowAnnotation("gene_anno" = anno_barplot(gene_anno, bar_width = 1, gp = gpar(fill = gene_anno_col), 
    width = anno_width), show_annotation_name = FALSE) +
rowAnnotation("cgi_anno" = anno_barplot(cgi_anno, bar_width = 1, gp = gpar(fill = cgi_anno_col), 
    width = anno_width), show_annotation_name = FALSE) +
Heatmap(mat_enrich_gf, name = "enrich_gf", col = z_score_col_fun, cluster_columns = FALSE,
    width = unit(ncol(mat_enrich_gf)*4, "mm"), column_title = "",
    heatmap_legend_param = list(title = "Z-score")) +
rowAnnotation("pct_st" = anno_barplot(mat_pct_st, bar_width = 1, gp = gpar(fill = state_col), 
    width = anno_width), show_annotation_name = FALSE) +
Heatmap(mat_enrich_st, name = "enrich_st", col = z_score_col_fun, cluster_columns = FALSE, 
    width = unit(ncol(mat_enrich_st)*6, "mm"), column_title = "", show_heatmap_legend = FALSE,
    column_names_gp = gpar(col = state_col), show_row_names = FALSE)

lgd_list = list(
    Legend(labels = c("gene", "intergenic"), title = "Gene annotation", 
        legend_gp = gpar(fill = gene_anno_col)),
    Legend(labels = c("<1kb", "1kb~5kb", "5kb~10kb", ">10kb"), title = "Distance to TSS", 
        legend_gp = gpar(fill = dist_tss_col)),
    Legend(labels = c("CGI", "CGI shore"), title = "CGI annotation", 
        legend_gp = gpar(fill = cgi_anno_col)),
    Legend(labels = colnames(mat_enrich_st), title = "Chromatin states", 
        legend_gp = gpar(fill = state_col))
)

draw(ht_list, padding = unit(c(2, 2, 20, 2), "mm"), row_split = gsub("\\d+$", "", label), 
    heatmap_legend_list = lgd_list, heatmap_legend_side = "bottom")
anno_title = c("n_gr" = "Number of\nDMRs", "n_corr" = "Significantly\ncorrelated genes",
    "gene_anno" = "Gene annotation", "dist_tss" = "Distance to TSS",
    "cgi_anno" = "CGI annotation", "pct_st" = "Overlap to\nChromatin states")
for(an in names(anno_title)) {
    decorate_annotation(an, {
        grid.text(anno_title[an], y = unit(1, "npc") + unit(3, "mm"), just = "bottom")
    })
}
ht_title = c("mean_meth" = "Mean\nmethylation", "enrich_gf" = "Enrichment to\ngenomic features",
    "enrich_st" = "Enrichment to\nchromatin states")
for(an in names(ht_title)) {
    decorate_heatmap_body(an, {
        grid.text(ht_title[an], y = unit(1, "npc") + unit(3, "mm"), just = "bottom")
    })
}
```

# Figure 6
```{r, message=FALSE, fig.width=12, fig.height=8, dev='svg'}

set.seed(123)
mat = matrix(rnorm(80, 2), 8, 10)
mat = rbind(mat, matrix(rnorm(40, -2), 4, 10))
rownames(mat) = paste0("R", 1:12)
colnames(mat) = paste0("C", 1:10)

ha_column1 = HeatmapAnnotation(points = anno_points(rnorm(10)), 
    annotation_name_side = "left")
ht1 = Heatmap(mat, name = "ht1", km = 2, column_title = "Heatmap 1", 
    top_annotation = ha_column1, row_names_side = "left", show_parent_dend_line = FALSE)

ha_column2 = HeatmapAnnotation(type = c(rep("a", 5), rep("b", 5)),
    col = list(type = c("a" = "red", "b" = "blue")))
ht2 = Heatmap(mat, name = "ht2", row_title = "Heatmap 2", column_title = "Heatmap 2", show_parent_dend_line = FALSE,
    bottom_annotation = ha_column2, column_km = 2)

ht_list = ht1 + ht2 + 
    rowAnnotation(bar = anno_barplot(rowMeans(mat), width = unit(2, "cm")))

ht_list = draw(ht_list, row_title = "Heatmap list", column_title = "Heatmap list", 
    heatmap_legend_side = "right", annotation_legend_side = "left")

# decorate_column_dend needs the following line in order to find the vairable in global scope
# decorate_dend doesn't need
# .GlobalEnv$ht1 <- ht1

decorate_dend("ht1", {
    tree = column_dend(ht1)
    ind = cutree(as.hclust(tree), k = 2)[order.dendrogram(tree)]

    first_index = function(l) which(l)[1]
    last_index = function(l) { x = which(l); x[length(x)] }
    x1 = c(first_index(ind == 1), first_index(ind == 2)) - 1
    x2 = c(last_index(ind == 1), last_index(ind == 2))
    grid.rect(x = x1/length(ind), width = (x2 - x1)/length(ind), just = "left",
        default.units = "npc", gp = gpar(fill = c("#FF000040", "#00FF0040"), col = NA))
    which = "column"
})

decorate_row_names("ht1", {
    grid.rect(gp = gpar(fill = "#FF000040"))
}, slice = 2)

decorate_row_title("ht1", {
    grid.rect(gp = gpar(fill = "#00FF0040"))
}, slice = 1)

decorate_annotation("points", {
    grid.lines(c(0, 1), unit(c(0, 0), "native"), gp = gpar(col = "red"))
})

sessionInfo()
```

	</div>
	<div class="splitter"></div>
	<div id="resultdiv" class="panel-right">
		<iframe id=result width=100% height=100% Color=result src=""></iframe>
	</div>
</div>
<script>
	$(".panel-left").resizable({

	   handleSelector: ".splitter",
	   resizeHeight:  false,
	   onDragEnd: function( event, ui ) {editor.resize();}
	});
</script>

</body>
</html>
